File: 2024/aoc-2024/src/aoc_2024/day09/day09.clj
(ns aoc-2024.day09.day09
  (:require [aoc-2024.utils.utils :as utils]))

(defn parse-input [raw_content]
  (let [raw_input (->> raw_content 
                       (seq) 
                       (map #(Integer/parseInt (str %1))))
        partitioned (partition 2 2 (list 0) raw_input)
        occupied (map first partitioned)
        free (map second partitioned)]
    [occupied free]))

(defn combine [v1 v2]
  (loop [l1 v1, l2 v2, out []]
    (cond
      (empty? l1) out
      :else (recur 
             (rest l1) 
             (rest l2) 
             (into out (concat (repeat (first l1) (- (count v1) (count l1))) 
                               (repeat (first l2) "#") 
                     ))))))

(defn compress [row]
  (loop [pending row, rev (reverse row), out '(), tail-to-ignore 0]
    (cond
      (empty? pending) (reverse out)
      (>= tail-to-ignore (count pending)) (reverse out)
      (= "#" (first rev)) (recur pending (rest rev) out (inc tail-to-ignore))
      (not= "#" (first pending)) (recur (rest pending) 
                                        rev 
                                        (conj out (first pending)) 
                                        tail-to-ignore)
      :else (recur (rest pending)
                   (rest rev)
                   (conj out (first rev))
                   (inc tail-to-ignore)))))

(defn checksum [input]
  (->> input 
       (map-indexed (fn [i x] (* i x )))
       (reduce +)))

(defn solve [[occupied free]]
  (let [combined (combine occupied free)
        compressed (compress combined)
        check (checksum compressed)]
    check))

(defn part1 
  ([] (part1 "day09/input.txt"))
  ([file-name]
  (-> file-name
      (utils/read-file)
      (parse-input)
      (solve))))

(defn count-first-n [coll target]
  (loop [v coll, out 0]
    (cond 
      (empty? v) out 
      (= (first v) target) (recur (rest v) (inc out)) 
      :else out)))

;; Should return (new-row, not-matched)
(defn compress-tail [row target n] 
  (loop [pending row, out []]
    (cond 
      (empty? pending) [(drop-last n row), (repeat n target)]
      (not= (first pending) "#") (recur (rest pending) (conj out (first pending)))
      ;; first = "#"
      :else (let [slots (count-first-n pending "#")]
              (if (>= slots n)
                [(concat out
                         (repeat n target)
                         (drop n (drop-last n pending))) []]
                (recur (drop slots pending) (concat out (take slots pending)) ))
              )
      )))

(defn compress2 [row]
  (let [[next-row unmatched] (compress-tail row
                                 (last row)
                                 (count-first-n (reverse row) (last row)))]
    (println "pre: " row )
    (println "post: " next-row )
    (println "unmatched: " unmatched "\n")

    (if (= row next-row)
      row
      (concat (compress2 next-row) unmatched))))
              
(defn solve2 [[occupied free]]
  (let [combined (combine occupied free)
        compressed (compress2 combined)
        k (println compressed)
        clean-compressed (map #(if (= %1 "#") 0 %1) compressed)
        k2 (println clean-compressed)
        check (checksum clean-compressed)]
    check))

(defn part2 
  ([] (part2 "day09/input.txt"))
  ([file-name]
  (-> file-name
      (utils/read-file)
      (parse-input)
      (solve2))))

(part2 "day09/ex1.txt")

(compress-tail '(0 0 "#" "#" "#" 1 1 "#" "#" 2 2) 2 4)

(concat [1 2 3] [4 5 6])

File: 2024/aoc-2024/src/aoc_2024/day09/day09.py
from itertools import zip_longest

def parse_input():
    raw_input = [int(x) for x in input()]
    occupied = raw_input[::2]
    free = raw_input[1::2]
    return (occupied, free)

def combine(occupied, free):
    resp = []

    zipped = zip_longest(occupied, free, fillvalue=0)

    for i, (oc, space) in enumerate(zipped):
        for j in range(oc):
            resp.append(i)
        for j in range(space):
            resp.append("#")
    return resp

def count_last_n(entry, target, right):
    resp = 0
    while entry[right] == target:
        resp += 1
        right -= 1

    return resp


def count_start_n(entry, target, left):
    resp = 0
    while entry[left] == target:
        resp += 1
        left +=1 
        
    return resp


def compress(entry):
    right = len(entry) - 1
    while right >= 0:
        last = entry[right]
        if last == "#":
            right -= 1
            continue

        last_freq = count_last_n(entry, last, right)

        update_index = -1
        should_update = False
        for left in range(right+1):
            if entry[left] == "#":
                available_pos = count_start_n(entry, "#", left)
                if available_pos >= last_freq:
                    should_update = True
                    update_index = left
                    break
        
        if should_update:
            for i in range(last_freq):
                entry[update_index + i] = last
                entry[right - i] = "#"
        right -= last_freq

    return entry


def checksum(entry):
    resp = 0
    for i, x in enumerate(entry):
        val = 0 if x == "#" else x
        resp += i * val
    return resp

def process(input):
    (occupied, free) = input

    combined = combine(occupied, free)
    compressed = compress(combined)
    
    return checksum(compressed)

def main():
    entry = parse_input()
    resp = process(entry)
    print(resp)

if __name__ == '__main__':
    main()

